lib/internal/async_hooks.js

```diff
@@ -337,13 +339,14 @@ function emitInitScript(asyncId, type, triggerAsyncId, resource) {
 }
 
 
-function emitBeforeScript(asyncId, triggerAsyncId) {
+function emitBeforeScript(asyncId, triggerAsyncId, resource) {
   // Validate the ids. An id of -1 means it was never set and is visible on the
   // call graph. An id < -1 should never happen in any circumstance. Throw
   // on user calls because async state should still be recoverable.
   validateAsyncId(asyncId, 'asyncId');
   validateAsyncId(triggerAsyncId, 'triggerAsyncId');
 
+  setCurrentResource(resource);
```

Comment:
assert here is that `currentResource === undefined || currentResource === null`.